### PR TITLE
Combine work-request errors into single reconcileError event.

### LINKED ISSUE #383: LoadBalancer/NLB work-request-errors could be included as part of events for better error information.
**What would you like to be added**:
Loadbalancer workRequestErrors are part of core services and cannot be fetched using workRequest. Due to this, the logs/events do not have the actual work-request errors and instead have this failure to fetch errors. This was encountered during testing of the PR: #378, which was merged before I could put the test-result for loadbalancer/networkloadbalancer.

**Why is this needed**:
The work-request-errors for loadbalancer and networkloadbalancer both should be populated in the events incase of failure of workrequests for better information.

### PR DESCRIPTION
**Pull request by sindhusri16**
Combine work-request errors into single reconcileError event so that processing of this single recent event would provide information about the status of the machine. This is only for ocimachine events and not nlb/lb

Related Fixes #383

### CODE REVIEW

**Comment by sindhusri16** [comment]
The change is only the second commit, the previous commit are already merged into main branch. The output of unittests:

> make test
fatal: No names found, cannot describe anything.
GOBIN=/home/opc/cluster-api-provider-oci/hack/tools/bin ./scripts/go_install.sh sigs.k8s.io/controller-runtime/tools/setup-envtest setup-envtest v0.0.0-20230131074648-f5014c077fc3
kube-builder assets: /home/opc/.local/share/kubebuilder-envtest/k8s/1.24.2-linux-amd64
KUBEBUILDER_ASSETS="/home/opc/.local/share/kubebuilder-envtest/k8s/1.24.2-linux-amd64" go test -coverprofile=coverage.out ./... 
?       github.com/oracle/cluster-api-provider-oci      [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/config [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/metrics        [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/scope/mocks    [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/base  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/base/mock_base        [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/compute       [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/compute/mock_compute  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement     [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement/mock_computemanagement      [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine       [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine/mock_containerengine  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/identity      [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/identity/mock_identity        [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/loadbalancer  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/loadbalancer/mock_lb  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer   [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer/mock_nlb  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/vcn   [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/vcn/mock_vcn  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/workrequests  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/workrequests/mock_workrequests        [no test files]
?       github.com/oracle/cluster-api-provider-oci/feature      [no test files]
?       github.com/oracle/cluster-api-provider-oci/version      [no test files]
ok      github.com/oracle/cluster-api-provider-oci/api/v1beta1  92.555s coverage: 23.4% of statements
ok      github.com/oracle/cluster-api-provider-oci/api/v1beta2  0.053s  coverage: 14.6% of statements
ok      github.com/oracle/cluster-api-provider-oci/cloud/ociutil        0.025s  coverage: 15.9% of statements
ok      github.com/oracle/cluster-api-provider-oci/cloud/scope  22.261s coverage: 76.1% of statements
ok      github.com/oracle/cluster-api-provider-oci/cloud/util   0.748s  coverage: 61.1% of statements
ok      github.com/oracle/cluster-api-provider-oci/controllers  2.559s  coverage: 59.2% of statements
ok      github.com/oracle/cluster-api-provider-oci/exp/api/v1beta1      32.768s coverage: 15.3% of statements
ok      github.com/oracle/cluster-api-provider-oci/exp/api/v1beta2      0.065s  coverage: 4.8% of statements
ok      github.com/oracle/cluster-api-provider-oci/exp/controllers      0.996s  coverage: 56.3% of statements
go tool cover -func=coverage.out -o coverage.txt
go tool cover -html=coverage.out -o coverage.html

File: controllers/ocimachine_controller.go [hunk 1]
```diff
@@ -316,12 +317,13 @@ func (r *OCIMachineReconciler) recordErrorsOnFailedWorkRequest(ctx context.Conte
 				r.Recorder.Event(machine, corev1.EventTypeWarning, "ReconcileError", errors.Wrapf(err, "Failed to normal reconcile OCIMachine").Error())
 				return reconcile.Result{}, errors.Wrapf(err, "failed to normal reconcile OCI Machine %s/%s", machineScope.OCIMachine.Namespace, machineScope.OCIMachine.Name)
 			}
+			final_error = final_error + fmt.Sprintf("Workrequest %s failed.", *wrqst.Id)
 			for _, wr_err := range wr_errs.Items {
-				r.Recorder.Eventf(machine, corev1.EventTypeWarning, "ReconcileError", *wr_err.Message)
+				final_error = final_error + " " + *wr_err.Message
```
Comment (joekr) [inline controllers/ocimachine_controller.go:322]: String formatting the Go way is typically using `fmt.Sprintf` like `fmt.Sprintf("%s %s", final_error, *wr_err.Message)` or whatever. I haven't tested that.

Comment (joekr) [reply controllers/ocimachine_controller.go:322]: Also by removing `r.Recorder.Eventf` isn't functionality lost? Isn't this actually recording an event here?

OH! this is the expected change? So this is just going to build a long error of all the possible error messages in the WR?

**Comment by joekr** [comment]
## e2e tests

```
Ran 7 of 28 Specs in 3084.062 seconds
SUCCESS! -- 7 Passed | 0 Failed | 0 Pending | 21 Skipped
```

**Review by joekr — APPROVED** [review]

**Pull request merged**